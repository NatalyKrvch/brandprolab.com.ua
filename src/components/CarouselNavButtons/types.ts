export interface CarouselNavButtonsProps {
  isFirstSlide: boolean;
  isLastSlide: boolean;
  onPreviousSlide: () => void;
  onNextSlide: () => void;
}
