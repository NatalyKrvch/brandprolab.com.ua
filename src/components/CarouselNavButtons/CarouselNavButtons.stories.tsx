import type { Meta, StoryObj } from '@storybook/react';

import CarouselNavButtons from './CarouselNavButtons';

const meta: Meta<typeof CarouselNavButtons> = {
  title: 'Components/CarouselNavButtons',
  component: CarouselNavButtons,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    isFirstSlide: {
      control: 'boolean',
      description: 'Whether the current slide is the first one.',
    },
    isLastSlide: {
      control: 'boolean',
      description: 'Whether the current slide is the last one.',
    },
    onPreviousSlide: {
      action: 'Previous clicked',
      description: 'Callback when previous button is clicked.',
    },
    onNextSlide: {
      action: 'Next clicked',
      description: 'Callback when next button is clicked.',
    },
  },
};

export default meta;
type Story = StoryObj<typeof CarouselNavButtons>;

const StoryWrapper = (args: any) => (
  <div className="relative h-[200px] w-[1200px] bg-gray-200">
    <CarouselNavButtons {...args} />
  </div>
);

export const Default: Story = {
  args: {
    isFirstSlide: false,
    isLastSlide: false,
  },
  render: StoryWrapper,
};

export const OnlyNextVisible: Story = {
  args: {
    isFirstSlide: true,
    isLastSlide: false,
  },
  render: StoryWrapper,
};

export const OnlyPreviousVisible: Story = {
  args: {
    isFirstSlide: false,
    isLastSlide: true,
  },
  render: StoryWrapper,
};

export const NoneVisible: Story = {
  args: {
    isFirstSlide: true,
    isLastSlide: true,
  },
  render: StoryWrapper,
};
