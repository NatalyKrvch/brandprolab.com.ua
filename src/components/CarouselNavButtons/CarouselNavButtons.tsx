import { AnimatePresence, motion } from 'framer-motion';
import Image from 'next/image';

import { CarouselNavButtonsProps } from './types';

const CarouselNavButtons = ({
  isFirstSlide,
  isLastSlide,
  onPrev,
  onNext,
}: CarouselNavButtonsProps) => {
  return (
    <div className="absolute right-0 hidden gap-2 tablet:-top-[60px] tablet:flex desktop:-top-[80px] desktop:flex">
      <AnimatePresence mode="wait">
        {!isFirstSlide && (
          <motion.button
            key="prev"
            onClick={onPrev}
            initial={{ opacity: 0, rotate: -15 }}
            animate={{ opacity: 1, rotate: 0 }}
            exit={{ opacity: 0, rotate: -15 }}
            transition={{ duration: 0.3 }}
          >
            <Image
              src="/icons/nav/row.svg"
              alt="Попередній слайд"
              width={0}
              height={0}
              className="h-[40px] w-[40px] rotate-180 desktop:h-[44px] desktop:w-[44px]"
            />
          </motion.button>
        )}

        {!isLastSlide && (
          <motion.button
            key="next"
            onClick={onNext}
            initial={{ opacity: 0, rotate: 15 }}
            animate={{ opacity: 1, rotate: 0 }}
            exit={{ opacity: 0, rotate: 15 }}
            transition={{ duration: 0.3 }}
          >
            <Image
              src="/icons/nav/row.svg"
              alt="Наступний слайд"
              width={0}
              height={0}
              className="h-[40px] w-[40px] desktop:h-[44px] desktop:w-[44px]"
            />
          </motion.button>
        )}
      </AnimatePresence>
    </div>
  );
};

export default CarouselNavButtons;
