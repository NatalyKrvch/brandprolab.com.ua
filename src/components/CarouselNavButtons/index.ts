export { default as CarouselNavButtons } from './CarouselNavButtons';
