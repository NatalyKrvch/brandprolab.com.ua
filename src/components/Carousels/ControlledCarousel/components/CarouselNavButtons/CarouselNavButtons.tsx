'use client';

import { AnimatePresence, motion } from 'framer-motion';
import Image from 'next/image';

import {
  CAROUSEL_NAV_NEXT_TEST_ID,
  CAROUSEL_NAV_PREV_TEST_ID,
  CAROUSEL_NAV_WRAPPER_TEST_ID,
} from '@/lib/testIDs';

import {
  fadeMotionProps,
  fadeNextButtonMotionProps,
} from './configs/motionConfig';
import type { CarouselNavButtonsProps } from './types';

const CarouselNavButtons = ({
  isFirstSlide,
  isLastSlide,
  onPreviousSlide,
  onNextSlide,
}: CarouselNavButtonsProps) => {
  return (
    <div
      role="group"
      aria-label="Навігація каруселі"
      className="absolute hidden gap-2 tablet:-top-80 tablet:right-16 tablet:flex desktop:-top-130 desktop:right-56 desktop:flex"
      data-testid={CAROUSEL_NAV_WRAPPER_TEST_ID}
    >
      <AnimatePresence>
        {!isFirstSlide && (
          <motion.button
            key="prev"
            aria-label="Попередній слайд"
            onClick={onPreviousSlide}
            {...fadeMotionProps}
            data-testid={CAROUSEL_NAV_PREV_TEST_ID}
          >
            <Image
              src="/icons/nav/row.svg"
              alt=""
              aria-hidden="true"
              width={40}
              height={40}
              className="h-40 w-40 rotate-180 desktop:h-44 desktop:w-44"
            />
          </motion.button>
        )}

        {!isLastSlide && (
          <motion.button
            key="next"
            aria-label="Наступний слайд"
            onClick={onNextSlide}
            {...fadeNextButtonMotionProps}
            data-testid={CAROUSEL_NAV_NEXT_TEST_ID}
          >
            <Image
              src="/icons/nav/row.svg"
              alt=""
              aria-hidden="true"
              width={40}
              height={40}
              className="h-40 w-40 desktop:h-44 desktop:w-44"
            />
          </motion.button>
        )}
      </AnimatePresence>
    </div>
  );
};

export default CarouselNavButtons;
