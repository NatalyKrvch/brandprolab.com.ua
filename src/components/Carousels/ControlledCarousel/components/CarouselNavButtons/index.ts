export { default as CarouselNavButtons } from './CarouselNavButtons';
export * from './types';
