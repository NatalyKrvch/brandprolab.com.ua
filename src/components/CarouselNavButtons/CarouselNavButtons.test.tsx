import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import CarouselNavButtons from './CarouselNavButtons';

describe('CarouselNavButtons component', () => {
  const defaultProps = {
    onPreviousSlide: jest.fn(),
    onNextSlide: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the previous button when not on the first slide', () => {
    render(
      <CarouselNavButtons
        {...defaultProps}
        isFirstSlide={false}
        isLastSlide={true}
      />,
    );

    const prevButton = screen.getByTestId('carousel-nav-prev-test-id');
    expect(prevButton).toBeInTheDocument();

    fireEvent.click(prevButton);
    expect(defaultProps.onPreviousSlide).toHaveBeenCalled();

    expect(document.body).toMatchSnapshot();
  });

  it('renders the next button when not on the last slide', () => {
    render(
      <CarouselNavButtons
        {...defaultProps}
        isFirstSlide={true}
        isLastSlide={false}
      />,
    );

    const nextButton = screen.getByTestId('carousel-nav-next-test-id');
    expect(nextButton).toBeInTheDocument();

    fireEvent.click(nextButton);
    expect(defaultProps.onNextSlide).toHaveBeenCalled();

    expect(document.body).toMatchSnapshot();
  });

  it('renders both buttons when not on first or last slide', () => {
    render(
      <CarouselNavButtons
        {...defaultProps}
        isFirstSlide={false}
        isLastSlide={false}
      />,
    );

    expect(screen.getByTestId('carousel-nav-prev-test-id')).toBeInTheDocument();
    expect(screen.getByTestId('carousel-nav-next-test-id')).toBeInTheDocument();

    expect(document.body).toMatchSnapshot();
  });

  it('renders no buttons when on first and last slide (edge case)', () => {
    render(
      <CarouselNavButtons
        {...defaultProps}
        isFirstSlide={true}
        isLastSlide={true}
      />,
    );

    expect(
      screen.queryByTestId('carousel-nav-prev-test-id'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('carousel-nav-next-test-id'),
    ).not.toBeInTheDocument();

    expect(document.body).toMatchSnapshot();
  });
});
