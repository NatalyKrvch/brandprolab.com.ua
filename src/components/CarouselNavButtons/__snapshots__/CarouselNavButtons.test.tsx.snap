// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselNavButtons component renders both buttons when not on first or last slide 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 hidden gap-2 tablet:-top-[60px] tablet:flex desktop:-top-[80px] desktop:flex"
      data-testid="carousel-nav-wrapper-test-id"
    >
      <button
        data-testid="carousel-nav-prev-test-id"
        style="opacity: 0; transform: rotate(-15deg);"
      >
        <img
          alt="Попередній слайд"
          class="h-[40px] w-[40px] rotate-180 desktop:h-[44px] desktop:w-[44px]"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/nav/row.svg"
          style="color: transparent;"
          width="0"
        />
      </button>
      <button
        data-testid="carousel-nav-next-test-id"
        style="opacity: 0; transform: rotate(15deg);"
      >
        <img
          alt="Наступний слайд"
          class="h-[40px] w-[40px] desktop:h-[44px] desktop:w-[44px]"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/nav/row.svg"
          style="color: transparent;"
          width="0"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`CarouselNavButtons component renders no buttons when on first and last slide (edge case) 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 hidden gap-2 tablet:-top-[60px] tablet:flex desktop:-top-[80px] desktop:flex"
      data-testid="carousel-nav-wrapper-test-id"
    />
  </div>
</body>
`;

exports[`CarouselNavButtons component renders the next button when not on the last slide 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 hidden gap-2 tablet:-top-[60px] tablet:flex desktop:-top-[80px] desktop:flex"
      data-testid="carousel-nav-wrapper-test-id"
    >
      <button
        data-testid="carousel-nav-next-test-id"
        style="opacity: 0; transform: rotate(15deg);"
      >
        <img
          alt="Наступний слайд"
          class="h-[40px] w-[40px] desktop:h-[44px] desktop:w-[44px]"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/nav/row.svg"
          style="color: transparent;"
          width="0"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`CarouselNavButtons component renders the previous button when not on the first slide 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 hidden gap-2 tablet:-top-[60px] tablet:flex desktop:-top-[80px] desktop:flex"
      data-testid="carousel-nav-wrapper-test-id"
    >
      <button
        data-testid="carousel-nav-prev-test-id"
        style="opacity: 0; transform: rotate(-15deg);"
      >
        <img
          alt="Попередній слайд"
          class="h-[40px] w-[40px] rotate-180 desktop:h-[44px] desktop:w-[44px]"
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          src="/icons/nav/row.svg"
          style="color: transparent;"
          width="0"
        />
      </button>
    </div>
  </div>
</body>
`;
